### ISSUE TITLE
Refactor components with options

### ISSUE THREAD

**Issue by HendrikThePendric** [dhis2/ui-core#662]
Currently `CheckboxGroup`, `RadioGroup` and `SwitchGroup` all assume that all options have a unique value. The internal logic in the components is based on that, and one of the consequences is that the selected options are simple represented as a string / an array of strings.

For the `SingleSelect` and `MultiSelect` it was decided that duplicate values should be allowed in  a list of options. ~So the internal logic is based around object reference equality, and the options are not simply strings, but objects.~ So every selected option is an object with a `value` and a `label` property and both are used to identify a unique option.

Since both of these types of components are essentially doing the same thing, they should probably be using the same assumptions in their implementation, so we should refactor things to make this equivalent.

The question is: which change are we going to be making here?

**Comment by HendrikThePendric** [comment]
My personal preference would be to just assume values are unique. This really simplifies things, and I think it's an assumption that will hold for 99% of cases. 

For the remaining 1% of cases, I think it should be possible to work around this, i.e. use some sort of suffix to force the value to be unique. What I think is important to realise for these type of scenarios, is that most likely some additional work will need to be done anyway (either before rendering or before posting), because you quite likely want to deduplicate values before posting them, etc. 

Also, the problem with supporting duplicate values is that you have to make assumptions about the desired behaviour in your implementation. Specifically, what do you show when an option is selected, that has a "twin" (i.e. another option with the same value)? Do you visually select the twin or not? I think the only valid answer would be "it depends". It simply depends on the context in which this is being used in. And we all know what happens when you start assuming.....

**Comment by ismay** [comment]
Small explanation:

> For the SingleSelect and MultiSelect it was decided that duplicate values should be allowed in a list of options. So the internal logic is based around object reference equality, and the options are not simply strings, but objects.

The Selects don't compare object references, but the actual contents (so whether value and label are the same), regardless of whether the reference is the same.

**Comment by HendrikThePendric** [comment]
I stand corrected. I'll update the issue description

**Comment by ismay** [comment]
Thanks. Yeah not sure if it's relevant, but just so that we're all on the same page with regard to the checks.

**Comment by ismay** [comment]
I think I can link to the original discussion on the Select and what type of options it should accept. Let me see if I can find it.

**Comment by ismay** [comment]
So the discussion on the option style starts here: https://github.com/dhis2/ui-core/pull/448#issuecomment-537110917

**Comment by ismay** [comment]
As for where to take the options. I'd be in favor of keeping the Select option style as-is. I think we had quite some discussion on it initially, and I feel like we arrived at what made the most sense.

As for the other components. I think consistency across all our components seems nice to have. Other than that I think that those discussions will probably mirror the discussion we had with the Select (assuming that native also behaves in a similar manner for the components as well).

**Comment by HendrikThePendric** [comment]
Perhaps also relevant in this discussion: https://dhis2.slack.com/archives/CBM8LNEQM/p1576061589056800

**TL;DR;**
_`label` was implemented as a prop, which isn't inline with the rest of ui-core (where we try to use children instead), and this was partly done to identify unique options easily._ 

So perhaps if we were to require unique option-values, we could refactor this and simply use children instead.

**Comment by ismay** [comment]
> Perhaps also relevant in this discussion: https://dhis2.slack.com/archives/CBM8LNEQM/p1576061589056800
> 
> **TL;DR;**
> _`label` was implemented as a prop in the Select, which isn't inline with the rest of ui-core (where we try to use children instead), and this was partly done to identify unique options easily._
> 
> So perhaps if we were to require unique option-values, we could refactor this and simply use children instead.

One thing where the label is also used by the way is the filter. Which filters on labels. The reason for not using children for the label is that it's not as clear that children have to be a label (string), and only that (so no nesting). Since this is a crucial part for the correct functioning of the Select we decided on a separate label prop, as that's more explicit.

**Comment by ismay** [comment]
I do feel like we've had a large part of this discussion already btw. We discussed the option style for the Select for a long time, weighed the options, arrived at a deliberate conclusion.

I do understand the need for consistency. But I do worry that we're about to start having the exact same discussion we had a while back.

**Comment by ismay** [comment]
We talked a bit about this on Slack today as well, here: https://dhis2.slack.com/archives/CBM8LNEQM/p1576174608112600. I tried to explain there what the thought process was for going with the current format for the Select:

> So, I see that it can be cumbersome. I think the choice we have is:
> 
> 1. Accept a single string value (or array of). This will not be able to deal with duplicate values, or labels.
> 2. Accept an object. It will be able to deal with duplicate values.
> 
> When discussing this we were looking at the native select. Which can deal with exact duplicates, as well as duplicate values. 
>
> We decided to compromise slightly, and just deal with duplicate values. Which led us to the current format.
>
> Realistically I think that in a lot of scenarios the dev will be mapping over data returned from the api, without ever knowing whether it contains any kind of duplicates or not. Regardless of whether we think it's a likely scenario or not. Our data seems to be very flexible, so I don't think it's impossible for it to occur.
>
> So if we allow for a single string value, it's quite likely that devs will be implementing it that way. That means that those selects will not be able to deal correctly with those types of values, if they occur. So in the end I think it's a question of whether we want to opt for a simpler API, with the chance of incorrectness at times, Or a more complex API that is correct more often.
>
> My vote has traditionally been on favouring correctness in the base design, because it seemed worse to me to allow for an incorrect UI, as opposed to a slightly more involved API. But that's my preference. If everyone else prefer the other option I'm all for it. I just think it's important to realise that this could cause bugs in the UI. And I don't think it's fair to say that that data pattern (with duplicate values) is an anti-pattern. It might reasonably occur, and a user doesn't know about our component requirements when configuring data, nor should they.
>
>I don't want it to seem like I have a vested interest in one format over the other btw. But it seems to me like going with the simpler format just places the constraints with our data, instead of the Select. And I don't think we're enforcing that our data be compliant with the format that the simpler Select would require?

Hope that this helps explain the thought process behind the current approach.

**Comment by ismay** [comment]
Hendrik and I talked about this some more last week. I've noticed that in the discussion of which option style to use, focus sometimes shifts from discussing the problem we're trying to solve for users (i.e. the requirements of our components), to discussing the preferred technical solution.

I think that the solution chosen for the Select followed from the requirements, the way I interpreted the requirements at the time, we were aiming for:

1. Safety (our components are used in applications that I presume should be as safe as possible, so I tended towards safety vs. a terse api). We already compromised here compared to a native select and chose the middle ground, inbetween string values and object reference equality.
2. A familiar api for users accustomed to html (using elements to supply options, instead of an object)
3. Customisability (allowing users to input custom options, and even invalid options for markup, inbetween headers, etc.)

If we want to revise the Select option style, I think it would be best to first discuss and establish the new requirements, instead of focusing on our preferred solution (i.e. strings for the new selected option format). The way I understand it, we are willing to compromise on point 1 for the sake of convenience (not judging, I get the motivation). Since our software deals with important information, I think it would be good to establish how this could fail in the case of unexpected data before we make any changes (we looked at this briefly in the past, and runtime checking seemed too costly). That way we can set clear requirements and discuss those before we start implementing.

**Comment by HendrikThePendric** [comment]
OK... I still stand by what I wrote in https://github.com/dhis2/ui-core/issues/662#issuecomment-564009418 but perhaps it would be good to provide some more details about the problem with the current approach as I have seen in both the ui-forms implementation and a PR in the usage-analytics-app, see https://github.com/dhis2/usage-analytics-app/pull/261#discussion_r367245610.

I will use the usage-analytics-app example as an illustration of the general problem:
https://github.com/dhis2/usage-analytics-app/blob/6ec25f3206fe911e3512b54c119bc57c8d0a5571/src/constants/intervals.js#L8-L13

- As you can see above, we had to construct a lookup structure for the options here.
- The reason for that was that in our form-state we didn't want to retain label/value data, but purely the selected values. I think this is true in virtually all cases.
- So we end up with a form-state that contains enough information to send requests to the server, but not enough information to control the Select's `selection`
- There would have been two options to solve this:
    1. Retain the superfluous data in the form state and write logic to pull the option value from the value/label object when constructing api requests.
    2. Work with an intermediate lookup object that can provide the value/label object to the select based on the value in the form-state (we went with this one)

As I see it, we have invented a hoop we will need to jump through virtually every time we use the select. The benefit of being able to deal with duplicate values, which I consider to be an exceptional case, simply doesn't outweigh the drawbacks we'd introduced in terms of inconvenience....

Especially given that fact that dealing with these duplicate values can be handled in the app too. Either by deduplicating or differentiating with value-suffices, whichever is appropriate....

**Comment by HendrikThePendric** [comment]
FYI, see https://github.com/dhis2/ui/pull/12/ for an implementation of a string based selection

**Comment by ismay** [comment]
We discussed this briefly on slack [here](https://dhis2.slack.com/archives/CBM8LNEQM/p1585554872076600). Quoting:

Goals for the option change:

> 1) api consistency with other form controls
> 2) developer experience
> 3) reducing internal complexityin order
>
> Considering how this is implemented in an app, the duplicate values would have to be de-duped and handled in the app for other purposes as well. So allowing duplicate values in the select when the app has to handle the complexity elsewhere anyway is another point i'm investigating

> we've discussed that we don't know what comes from the API, that it may contain duplicate values, so if we pipe that data straight through to the select, then it may break
> but we didn't discuss the reality of that passthrough -- i think that the app would have to handle the data beforehand anyway, and in that process it needs to deal with duplicate values as well.
> but this is a hypothetical point so far, i haven't found any instances of duplicate values in the implementations of the select so far
> which in turn leads me to believe that we have solved a problem that in practice doesn't exist for most uses
> so a simpler api would benefit current implementations in data-viz and usage analytics, et. al.

So if I'm not mistaken, @varl is still investigating, but it does seem like we don't have duplicate values in our select implementations so far. Plus it seems like the apps are already dealing with that themselves anyway.

In which case I agree with @HendrikThePendric and @varl, that the simpler API outweighs the safety benefits. It doesn't seem like we need them (the safety benefits), and the simpler API would be less cumbersome to use.

**Comment by varl** [comment]
Thanks for quoting in the relevant parts of my Slack rant, @ismay.

I've gone over our current implementations of the Select and I cannot find any instances of duplicate values. Often we are pointing straight to a unique identifier we get from the backend as the value. It's not the only case, though I cannot find duplicate values used in e.g. Usage Analytics or Data Visualizer either.

Since we used MUI historically, I wasn't too surprised as they also use strings for values. However, they also support using an Object as a value, which I found interesting:

> Set to an empty string '' if you don't want any of the available options to be selected.
**If the value is an object it must have reference equality with the option in order to be selected.** If the value is not an object, the string representation must match with the string representation of the option in order to be selected.[1]

This is perhaps something we could trivially support as well.

[1] https://material-ui.com/api/select/

**Comment by HendrikThePendric** [comment]
> This is perhaps something we could trivially support as well.

Could you provide some clarification reg. the above @varl? Specifically, I'd like to know whether you'd want to add support for both strings and objects right now, or whether you are just discussing steps we could possibly take in the future.

I seem to remember from our talk last week that it was your preference to go with a singular approach, and that's the way I've implemented things in dhis2/ui#12.

Looking at the code-changes in that PR, we see that adding support for both objects and strings would be fairly trivial in terms of the actual implementation. However, we'd have to add quite a few tests to cover this "dual functionality", so in the end it'd be a fairly substantial task.

My personal preference would be too go with this purely string-based approach for now and (re) add support for objects when the need presents itself. It wouldn't produce breaking changes.

> If the value is an object it must have reference equality with the option in order to be selected.

One note about this: our solution wouldn't work that way, instead we'd individually check the `label` and `value` properties by value. This is cause we're pulling the value and label off the option since we're working with children and not an array of options.

**Comment by varl** [comment]
>Specifically, I'd like to know whether you'd want to add support for both strings and objects right now, or whether you are just discussing steps we could possibly take in the future.

I meant it as a path we could take in the future if we need it, I don't think we want it now.

**Comment by ismay** [comment]
I wonder how they mean this exactly. They don't allow passing options as an object or an array of objects. I also couldn't find an example in their demos of passing an object to the `value` prop. So my question would be, reference equality to which object exactly?

**Comment by varl** [comment]
The Option `value` and Select `value` need to match: https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Select/Select.test.js#L281-L292

**Comment by ismay** [comment]
Ah. So it's not actually reference equality, but a deep compare.

**Comment by ismay** [comment]
Ah ok, so you can pass an object to an option, as well as a string. Could be actually that we unintentionally already sort of support this with the comparisons we're doing

**Comment by ismay** [comment]
But yeah I think the direction we chose with accepting strings for `selected` is a good balance of tradeoffs. Maybe that balance will change in the future with new information, but I think I'd also favor a singular, string based approach for now, based on our current information.

**Comment by varl** [comment]
Done in https://github.com/dhis2/ui/pull/12. Closing this as a result.
